### ISSUE TITLE
BUG: MultiIndex.unique incorrect when NA value is present

### ISSUE THREAD

**Issue by jbrockmendel** [pandas-dev/pandas#41823]
drop_duplicates looks right, while unique (and _get_unique_index) keep a duplicate entry containing an NA value.  Example based on test_intersection_with_missing_values_on_both_sides.  The analogous problem crops up for every entry in nulls_fixture

```
nulls_fixture = np.nan
mi1 = pd.MultiIndex.from_arrays([[3, nulls_fixture, 4, nulls_fixture], [1, 2, 4, 2]])

>>> mi1.unique()
MultiIndex([(3.0, 1),
            (nan, 2),
            (4.0, 4),
            (nan, 2)],
           )

>>> mi1.drop_duplicates()
MultiIndex([(3.0, 1),
            (nan, 2),
            (4.0, 4)],
           )
```

**Comment by jbrockmendel** [comment]
@realead i think one of your recent PRs closes this, not sure which one

**Comment by realead** [comment]
It was closed with #41952

**Comment by realead** [comment]
The difference is , that `mi1.unique` operates on `self.values`  (https://github.com/pandas-dev/pandas/blob/648eb40abc53bcbdb2a6a1e68fd309c3d8021f8a/pandas/core/base.py#L969)

(i.e.  `[(3.0, 1) (nan, 2) (4.0, 4) (nan, 2)]`) , while `drop_duplicates` operates on `self.codes` https://github.com/pandas-dev/pandas/blob/648eb40abc53bcbdb2a6a1e68fd309c3d8021f8a/pandas/core/indexes/multi.py#L1592

where nans are replaced by -1.

This explains while `drop_duplicates` worked, but not  `unique`.
